apps/utils/cipher/cipher.main.kts

```diff
@@ -0,0 +1,462 @@
+
+
+val APP_NAME = "cipher"
+val DEFAULT_REGION = "us-east-1"
+val DEFAULT_EDITOR = "vi"
+
+enum class Token(val bytes: ByteString) {
+    SecureStart("<<SECURE>>".encodeUtf8()),
+    SecureEnd("<</SECURE>>".encodeUtf8()),
+    CipherStart("<<CIPHER>>".encodeUtf8()),
+    CipherEnd("<</CIPHER>>".encodeUtf8()),
+    None(ByteString.of());
+
+    val endsWith: Token? by lazy {
+        when (this) {
+            SecureStart -> SecureEnd
+            CipherStart -> CipherEnd
+            else -> null
+        }
+    }
+
+    val text: String by lazy {
+        bytes.utf8()
+    }
+
+    val isEnd: Boolean by lazy {
+        when (this) {
+            SecureEnd -> true
+            CipherEnd -> true
+            else -> false
+        }
+    }
+}
+
+enum class Span(val start: Token, val end: Token) {
+    Secure(Token.SecureStart, Token.SecureEnd),
+    Cipher(Token.CipherStart, Token.CipherEnd),
+    Text(Token.None, Token.None)
+}
+
+data class Block(val endToken: Token, val bytes: ByteString) {
+    override fun toString(): String = "($endToken: ${bytes.utf8()})"
+}
+
+class UsageException(message: String) : Exception(message)
+
+class ParseException(message: String) : Exception(message) {
+    companion object {
+        fun noStartToken(token: Token) = ParseException("end token has no start token: ${token.text}")
+        fun noEndToken(token: Token) = ParseException("start token has no end token: ${token.text}")
+    }
+}
+
+class CipherCache private constructor(private val map: PersistentMap<String, ByteString>) {
+    constructor() : this(persistentMapOf())
+
+    fun put(context: ByteString, secureText: ByteString, cipherText: ByteString) =
+        CipherCache(map.put(makeKey(context, secureText), cipherText))
+
+    fun get(context: ByteString, secureText: ByteString): ByteString? = map[makeKey(context, secureText)]
+
+    private fun makeKey(context: ByteString, secureText: ByteString): String {
+        val newLineString = "\n".encodeUtf8()
+        val source = context.toBuffer()
+        var newLine = source.indexOf(newLineString)
+        while (newLine >= 0) {
+            source.skip(newLine + newLineString.size)
+            newLine = source.indexOf(newLineString)
+        }
+        val keyPrefix = source.readUtf8()
+        val keySuffix = secureText.utf8()
+        return "$keyPrefix:$keySuffix"
+    }
+}
+
+fun ByteString.toBuffer(): BufferedSource =
+    Buffer().readFrom(ByteArrayInputStream(toByteArray()))
+
+fun BufferedSource.nextBlock(): Block? {
+    if (this.exhausted()) {
+        return null
+    }
+
+    var firstIndex = -1L
+    var firstToken = Token.None
+    for (token in Token.values()) {
+        if (token.bytes.size > 0) {
+            val index = this.indexOf(token.bytes)
+            if (index >= 0) {
+                if (firstIndex < 0 || index < firstIndex) {
+                    firstIndex = index
+                    firstToken = token
+//                println("$index $token")
```

Comment:
Should these commented lines be removed?

Reply:
Since this is a script and these are helpful for debugging I left them in rather than have to recreate them later if we need to debug some more.